// devexpress-reporting/viewer/internal/_progressViewModel.js
/**
* DevExpress HTML/JS Reporting (viewer\internal\_progressViewModel.js)
* Version:  20.2.4
* Build date: Dec 2, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var ko = require("knockout");
var $ = require("jquery");
var ProgressViewModel = (function () {
    function ProgressViewModel() {
        var _this = this;
        this.progress = ko.observable(0);
        this._forceInvisible = ko.observable(false);
        this.inProgress = ko.observable(false);
        this.cancelText = ko.observable(analytics_utils_1.getLocalization('Cancel', 'AnalyticsCoreStringId.SearchDialog_Cancel'));
        this.text = ko.observable('');
        this.cssClasses = ko.observable();
        this.visible = ko.pureComputed({
            read: function () {
                if (_this._forceInvisible()) {
                    return false;
                }
                return _this.inProgress();
            },
            write: function (visibleState) {
                _this._forceInvisible(!visibleState);
            }
        });
        this.complete = function () {
            _this.inProgress(false);
            _this.progress(0);
            $.isFunction(_this._onComplete) && _this._onComplete();
            _this._onComplete = null;
        };
        this.startProgress = function (onComplete, onStop) {
            _this.inProgress(true);
            _this.progress(0);
            _this._onComplete = onComplete;
            _this.stop = function () {
                try {
                    $.isFunction(onStop) && onStop();
                }
                finally {
                    _this.complete();
                }
            };
        };
    }
    ProgressViewModel.prototype.setPosition = function (position) {
        if (!position) {
            this.cssClasses({
                'dxrd-align-default': true
            });
        }
        else {
            this.cssClasses({
                'dxrd-align-bottom': position.bottom,
                'dxrd-align-top': position.top,
                'dxrd-align-left': position.left,
                'dxrd-align-right': position.right
            });
        }
    };
    return ProgressViewModel;
}());
exports.ProgressViewModel = ProgressViewModel;
