// @devexpress/analytics-core/widgets/common/pathRequest.js
/**
* DevExpress Analytics (widgets\common\pathRequest.js)
* Version:  20.2.4
* Build date: Dec 2, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PathRequest = (function () {
    function PathRequest(fullPath, pathParts) {
        if (pathParts === void 0) { pathParts = []; }
        this.pathParts = pathParts;
        this.path = '';
        this.fullPath = fullPath;
        if (fullPath) {
            if (fullPath.indexOf('.') !== -1) {
                var pathComponents = fullPath.split('.');
                this.id = this.ref = pathComponents[0];
                pathComponents.splice(0, 1);
                this.path = pathComponents.join('.');
            }
            else {
                this.id = this.ref = fullPath;
            }
        }
    }
    return PathRequest;
}());
exports.PathRequest = PathRequest;
